### PR TITLE
Add unit test for `Node3D`

### PR DESCRIPTION
**Pull request by JayTropper**
This commit adds several test cases for the Node3D class for #43440.

The test consists of two basic test cases and nine test cases that cover several utility methods like `look_at`, `rotate_object_local`, `to_local`and more.

### CODE REVIEW

File: tests/test_main.cpp [hunk 1]
```diff
@@ -118,6 +118,7 @@
 #include "tests/scene/test_instance_placeholder.h"
 #include "tests/scene/test_node.h"
 #include "tests/scene/test_node_2d.h"
+#include "tests/scene/test_node_3d.h"
```
Comment (AThousandShips) [inline tests/test_main.cpp:121]: Should be moved below to `#ifndef _3D_DISABLED`

**Comment by AThousandShips** [comment]
The commit is associated with another user, you will need to reset the author with `git commit --reset-author --amend`

File: tests/scene/test_node_3d.h [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		test_node3->look_at(target_node->get_global_position());
+		CHECK(test_node3->get_global_rotation().is_equal_approx(Vector3(0.093183, 1.086318, 0)));
+
+		test_node3->set_rotation(Vector3(0, 0, 0));
+		memdelete(target_node);
+	}
+
+	SUBCASE("[Node3D] rotate_object_local") {
+		// Basic rotate_object_local tests
+		test_node3->rotate_object_local(Vector3(0, 1, 0), real_t(Math::deg_to_rad(180.00)));
```
Comment (AThousandShips) [inline tests/scene/test_node_3d.h:148]:
```suggestion
		test_node3->rotate_object_local(Vector3(0, 1, 0), Math_PI);
```
Since these methods use radians it'd be best to remove an error factor based on conversion IMO

File: tests/scene/test_node_3d.h [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+		CHECK(test_node3->get_global_rotation_degrees().is_equal_approx(Vector3(0, -180, 0)));
+		CHECK(test_node3->get_position().is_equal_approx(Vector3(0, 0, 10)));
+		CHECK(test_node3->get_rotation_degrees().is_equal_approx(Vector3(0, -180, 0)));
+		CHECK(test_node3->get_scale().is_equal_approx(Vector3(2, 2, 2)));
+		CHECK_EQ(test_node3->get_rotation_edit_mode(), Node3D::ROTATION_EDIT_MODE_EULER);
+		CHECK_EQ(test_node3->get_rotation_order(), EulerOrder::YXZ);
+
+		// Change rotation order and rotate again
+		test_node3->set_rotation_order(EulerOrder::XYZ);
+		test_node3->rotate_object_local(Vector3(0, 0, -1), real_t(Math::deg_to_rad(90.00)));
```
Comment (AThousandShips) [inline tests/scene/test_node_3d.h:159]:
```suggestion
		test_node3->rotate_object_local(Vector3(0, 0, -1), Math_PI / 2.0);
```

File: tests/scene/test_node_3d.h [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+		CHECK(test_node3->get_global_position().is_equal_approx(Vector3(200, 0, 110)));
+		CHECK(test_node3->get_global_rotation_degrees().is_equal_approx(Vector3(0, -180, -90)));
+		CHECK(test_node3->get_position().is_equal_approx(Vector3(0, 0, 10)));
+		CHECK(test_node3->get_rotation_degrees().is_equal_approx(Vector3(-180, -0.000009, 90)));
+		CHECK(test_node3->get_scale().is_equal_approx(Vector3(2, 2, 2)));
+		CHECK_EQ(test_node3->get_rotation_order(), EulerOrder::XYZ);
+
+		// Do a rotation on several axes
+		test_node3->rotate_object_local(Vector3(-1, -1, -1).normalized(), real_t(Math::deg_to_rad(270.00)));
```
Comment (AThousandShips) [inline tests/scene/test_node_3d.h:168]:
```suggestion
		test_node3->rotate_object_local(Vector3(-1, -1, -1).normalized(), 1.5 * Math_PI);
```

File: tests/scene/test_node_3d.h (hunk 2 above)
Comment (JayTropper) [reply tests/scene/test_node_3d.h:148]: Yes, that is much smoother.

**Comment by AThousandShips** [comment]
If the tests still fail on the last build it will be because you generated your expected values on a non-doubles build, then you'll need to build a double precision build (or print the values in the test run) and update them, so they work in both cases (the double precision values will work for single precision)

File: tests/scene/test_node_3d.h [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+Vector3 wrap_rotation_angles(Vector3 rotation) {
+	// The first if check is necessary for angles around 0, especially ones slightly below 0,
+	//   e.g. 0.001 would stay the same and -0.001 would change to 359.999, which can lead to
+	//   other values when changing the build precision.
+	if (rotation.is_equal_approx(Vector3(0, rotation.y, rotation.z))) {
+		rotation.x = 0;
+	} else if (rotation.x < 0) {
+		rotation.x += 360;
+	}
+	if (rotation.is_equal_approx(Vector3(rotation.x, 0, rotation.z))) {
```
Comment (AThousandShips) [inline tests/scene/test_node_3d.h:49]:
```suggestion
	if (Math.is_zero_approx(rotation.y)) {
```

File: tests/scene/test_node_3d.h [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+		rotation.y = 0;
+	} else if (rotation.y < 0) {
+		rotation.y += 360;
+	}
+	if (rotation.is_equal_approx(Vector3(rotation.x, rotation.y, 0))) {
```
Comment (AThousandShips) [inline tests/scene/test_node_3d.h:54]:
```suggestion
	if (Math.is_zero_approx(rotation.z)) {
```

File: tests/scene/test_node_3d.h [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+#include "scene/main/window.h"
+#include "tests/test_macros.h"
+
+namespace TestNode3D {
+
+Vector3 wrap_rotation_angles(Vector3 rotation) {
+	// The first if check is necessary for angles around 0, especially ones slightly below 0,
+	//   e.g. 0.001 would stay the same and -0.001 would change to 359.999, which can lead to
+	//   other values when changing the build precision.
+	if (rotation.is_equal_approx(Vector3(0, rotation.y, rotation.z))) {
```
[the remark this answers is not in the document]
Comment (akien-mga) [reply tests/scene/test_node_3d.h:44]: Note it should be a colon in `Math::is_zero_approx`.

File: tests/scene/test_node_3d.h [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+		CHECK_EQ(test_node3->get_rotation_edit_mode(), Node3D::ROTATION_EDIT_MODE_EULER);
+		CHECK_EQ(test_node3->get_rotation_order(), EulerOrder::YXZ);
+
+		// Change rotation order and rotate again
+		test_node3->set_rotation_order(EulerOrder::XYZ);
+		test_node3->rotate_object_local(Vector3(0, 0, -1), Math_PI / 2.0);
+		CHECK(test_node3->get_global_position().is_equal_approx(Vector3(200, 0, 110)));
+		CHECK(wrap_rotation_angles(test_node3->get_global_rotation_degrees()).is_equal_approx(Vector3(0, 180, 270)));
+		CHECK(test_node3->get_position().is_equal_approx(Vector3(0, 0, 10)));
+		CHECK(wrap_rotation_angles(test_node3->get_rotation_degrees()).is_equal_approx(Vector3(180, 0, 90)));
```
Comment (JayTropper) [inline tests/scene/test_node_3d.h:182]: Unfortunately, `posmod(360)` only works in the main cases and not in the corner case I have in this line of code. 
The corner case: When executing with floats, the value of the y axes is -0.000009. When executing with doubles, the value of the y axes is 0.00000000000001. That means in the case of posmod(360) for double, no modulo is applied and the value stays the same. In the case of float precision, since it has a negative value, the value is adapted to 359,999991, which will lead to a failing test result since the test checks for 0. I hope that makes the situation clear.
We can leave it like that, or I could remove this case and take other values to avoid this corner case. Then there is no need for the `wrap_rotation_angles()` method since it only appears in this specific case.

Comment (JayTropper) [reply tests/scene/test_node_3d.h:182]: @AThousandShips @akien-mga Should I just remove this case or change the values?

Comment (Calinou) [reply tests/scene/test_node_3d.h:182]: I would remove the test case here for now. We can reintroduce it in a future PR.

File: tests/scene/test_node_3d.h [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+		test_node3->look_at(target_node->get_global_position());
+		CHECK(test_node3->get_global_rotation().is_equal_approx(Vector3(0.093183, 1.086318, 0)));
+
+		test_node3->set_rotation(Vector3(0, 0, 0));
+		memdelete(target_node);
+	}
+
+	SUBCASE("[Node3D] rotate_object_local") {
+		// Basic rotate_object_local tests
+		test_node3->rotate_object_local(Vector3(0, 1, 0), Math_PI);
```
Comment (Calinou) [inline tests/scene/test_node_3d.h:167]:
```suggestion
		test_node3->rotate_object_local(Vector3(0, 1, 0), Math::PI);
```

File: tests/scene/test_node_3d.h [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+		CHECK(wrap_rotation_angles(test_node3->get_global_rotation_degrees()).is_equal_approx(Vector3(0, 180, 0)));
+		CHECK(test_node3->get_position().is_equal_approx(Vector3(0, 0, 10)));
+		CHECK(wrap_rotation_angles(test_node3->get_rotation_degrees()).is_equal_approx(Vector3(0, 180, 0)));
+		CHECK(test_node3->get_scale().is_equal_approx(Vector3(2, 2, 2)));
+		CHECK_EQ(test_node3->get_rotation_edit_mode(), Node3D::ROTATION_EDIT_MODE_EULER);
+		CHECK_EQ(test_node3->get_rotation_order(), EulerOrder::YXZ);
+
+		// Change rotation order and rotate again
+		test_node3->set_rotation_order(EulerOrder::XYZ);
+		test_node3->rotate_object_local(Vector3(0, 0, -1), Math_PI / 2.0);
```
Comment (Calinou) [inline tests/scene/test_node_3d.h:178]:
```suggestion
		test_node3->rotate_object_local(Vector3(0, 0, -1), Math::PI / 2.0);
```

File: tests/scene/test_node_3d.h [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+		CHECK(test_node3->get_global_position().is_equal_approx(Vector3(200, 0, 110)));
+		CHECK(wrap_rotation_angles(test_node3->get_global_rotation_degrees()).is_equal_approx(Vector3(0, 180, 270)));
+		CHECK(test_node3->get_position().is_equal_approx(Vector3(0, 0, 10)));
+		CHECK(wrap_rotation_angles(test_node3->get_rotation_degrees()).is_equal_approx(Vector3(180, 0, 90)));
+		CHECK(test_node3->get_scale().is_equal_approx(Vector3(2, 2, 2)));
+		CHECK_EQ(test_node3->get_rotation_order(), EulerOrder::XYZ);
+
+		// Do a rotation on several axes
+		test_node3->rotate_object_local(Vector3(-1, -1, -1).normalized(), Math_PI * 1.5);
```
Comment (Calinou) [inline tests/scene/test_node_3d.h:187]:
```suggestion
		test_node3->rotate_object_local(Vector3(-1, -1, -1).normalized(), Math::PI * 1.5);
```

**Review by Calinou — COMMENTED** [review]
Tested locally on Linux (rebased on top of `master` https://github.com/godotengine/godot/commit/9dde5688a566775df338df8ae7023cb69b6cfe6d), it works as expected. Code looks good to me.

In addition to fixing merge conflicts, please squash the commits together so we can merge this. See [PR workflow](https://contributing.godotengine.org/en/latest/organization/pull_requests/creating_pull_requests.html#the-interactive-rebase) for instructions :slightly_smiling_face:
